Add mypy

Mypy turned off for initial development. We should add it again.

Main issue is currently in passing the configuration values to the Core object:

```
deptry/cli.py:67: error: Argument "ignore_dependencies" to "Core" has incompatible type "Union[bool, List[str], None]"; expected "List[str]"  [arg-type]
deptry/cli.py:68: error: Argument "ignore_directories" to "Core" has incompatible type "Union[bool, List[str], None]"; expected "List[str]"  [arg-type]
deptry/cli.py:69: error: Argument "ignore_notebooks" to "Core" has incompatible type "Union[bool, List[str], None]"; expected "bool"  [arg-type]
```

--- a/deptry/cli.py
+++ b/deptry/cli.py
@@ -6,9 +6,10 @@ from typing import List, Optional, Tuple, Union
 import click
 
 from deptry.cli_defaults import DEFAULTS
+from deptry.compat import metadata
 from deptry.config import read_configuration_from_pyproject_toml
 from deptry.core import Core
-from deptry.utils import PYPROJECT_TOML_PATH, import_importlib_metadata, run_within_dir
+from deptry.utils import PYPROJECT_TOML_PATH, run_within_dir
 
 
 class CommaSeparatedTupleParamType(click.ParamType):
@@ -226,5 +227,4 @@ def deptry(
 
 
 def display_deptry_version() -> None:
-    metadata, *_ = import_importlib_metadata()
     logging.info(f'deptry {metadata.version("deptry")}')

--- a/deptry/compat.py
+++ b/deptry/compat.py
@@ -1,0 +1,8 @@
+import sys
+
+if sys.version_info >= (3, 8):
+    import importlib.metadata as metadata
+    from importlib.metadata import PackageNotFoundError
+else:
+    import importlib_metadata as metadata  # noqa: F401
+    from importlib_metadata import PackageNotFoundError  # noqa: F401

--- a/deptry/core.py
+++ b/deptry/core.py
@@ -47,8 +47,10 @@ class Core:
             exclude=self.exclude + self.extend_exclude, ignore_notebooks=self.ignore_notebooks
         ).get_all_python_files_in(Path("."))
 
-        imported_modules = ImportParser().get_imported_modules_for_list_of_files(all_python_files)
-        imported_modules = [ModuleBuilder(mod, dependencies, dev_dependencies).build() for mod in imported_modules]
+        imported_modules = [
+            ModuleBuilder(mod, dependencies, dev_dependencies).build()
+            for mod in ImportParser().get_imported_modules_for_list_of_files(all_python_files)
+        ]
         imported_modules = [mod for mod in imported_modules if not mod.standard_library]
 
         issues = self._find_issues(imported_modules, dependencies, dev_dependencies)
@@ -106,7 +108,7 @@ class Core:
         logging.debug("")
 
     @staticmethod
-    def _exit(issues):
+    def _exit(issues: Dict[str, List[str]]) -> None:
         total_issues_found = sum([len(v) for k, v in issues.items()])
         if total_issues_found > 0:
             sys.exit(1)

--- a/deptry/dependency.py
+++ b/deptry/dependency.py
@@ -2,9 +2,7 @@ import logging
 import re
 from typing import List, Set
 
-from deptry.utils import import_importlib_metadata
-
-metadata, PackageNotFoundError = import_importlib_metadata()
+from deptry.compat import PackageNotFoundError, metadata
 
 
 class Dependency:

--- a/deptry/dependency.py
+++ b/deptry/dependency.py
@@ -100,11 +100,16 @@ class Dependency:
         """
         top_levels = []
         try:
-            metadata_records = metadata.distribution(self.name).read_text("RECORD").split("\n")
-        except:  # noqa: E722
+            metadata_records = metadata.distribution(self.name).read_text("RECORD")
+
+            if not metadata_records:
+                return []
+        except Exception:
             return []
-        for line in metadata_records:
+
+        for line in metadata_records.split("\n"):
             match = re.match("([a-zA-Z0-9-_]+)/", line)
             if match:
                 top_levels.append(match.group(1))
+
         return list(set(top_levels))

--- a/deptry/dependency_getter/requirements_txt.py
+++ b/deptry/dependency_getter/requirements_txt.py
@@ -1,7 +1,7 @@
 import logging
 import os
 import re
-from typing import List, Tuple, Union
+from typing import List, Optional, Tuple, Union
 
 from deptry.dependency import Dependency
 
@@ -112,7 +112,7 @@ class RequirementsTxtDependencyGetter:
         logging.debug("")
 
     @staticmethod
-    def _line_is_url(line: str):
+    def _line_is_url(line: str) -> Optional[re.Match]:
         return re.search("^(http|https|git\+https)", line)
 
     @staticmethod

--- a/deptry/dependency_specification_detector.py
+++ b/deptry/dependency_specification_detector.py
@@ -55,8 +55,9 @@ class DependencySpecificationDetector:
                 logging.debug(
                     "pyproject.toml does not contain a [tool.poetry.dependencies] section, so it is not used to specify the project's dependencies."
                 )
-                return False
-            return None
+                pass
+
+        return False
 
     def _check_if_project_uses_requirements_txt_for_dependencies(self) -> bool:
         if os.path.isfile(self.requirements_txt):

--- a/deptry/issue_finders/misplaced_dev.py
+++ b/deptry/issue_finders/misplaced_dev.py
@@ -1,5 +1,5 @@
 import logging
-from typing import List, Tuple
+from typing import List, Optional, Tuple
 
 from deptry.dependency import Dependency
 from deptry.module import Module
@@ -54,7 +54,7 @@ class MisplacedDevDependenciesFinder:
                 return True
         return False
 
-    def _get_package_name(self, module: Module):
+    def _get_package_name(self, module: Module) -> Optional[str]:
         if module.package:
             return module.package
         if module.dev_top_levels:

--- a/deptry/issue_finders/transitive.py
+++ b/deptry/issue_finders/transitive.py
@@ -1,5 +1,5 @@
 import logging
-from typing import List, Tuple
+from typing import List, Optional, Tuple
 
 from deptry.dependency import Dependency
 from deptry.module import Module
@@ -24,7 +24,7 @@ class TransitiveDependenciesFinder:
         self.dependencies = dependencies
         self.ignore_transitive = ignore_transitive
 
-    def find(self) -> List[str]:
+    def find(self) -> List[Optional[str]]:
         logging.debug("\nScanning for transitive dependencies...")
         transitive_dependencies = []
         for module in self.imported_modules:

--- a/deptry/json_writer.py
+++ b/deptry/json_writer.py
@@ -13,6 +13,6 @@ class JsonWriter:
     def __init__(self, json_output: str) -> None:
         self.json_output = json_output
 
-    def write(self, issues: Dict):
+    def write(self, issues: Dict) -> None:
         with open(self.json_output, "w", encoding="utf-8") as f:
             json.dump(issues, f, ensure_ascii=False, indent=4)

--- a/deptry/module.py
+++ b/deptry/module.py
@@ -55,7 +55,7 @@ class ModuleBuilder:
         Args:
             name: The name of the imported module
             dependencies: A list of the project's dependencies
-            dependencies: A list of the project's development dependencies
+            dev-dependencies: A list of the project's development dependencies
         """
         self.name = name
         self.dependencies = dependencies
@@ -90,14 +90,14 @@ class ModuleBuilder:
             is_dev_dependency=is_dev_dependency,
         )
 
-    def _get_package_name_from_metadata(self) -> str:
+    def _get_package_name_from_metadata(self) -> Optional[str]:
         """
         Most packages simply have a field called "Name" in their metadata. This method extracts that field.
         """
         try:
             return metadata.metadata(self.name)["Name"]
         except PackageNotFoundError:
-            pass
+            return None
 
     def _get_corresponding_top_levels_from(self, dependencies: List[Dependency]) -> List[str]:
         """
@@ -144,15 +144,15 @@ class ModuleBuilder:
         local_modules = [subdir for subdir in directories if "__init__.py" in os.listdir(subdir)]
         return self.name in local_modules
 
-    def _has_matching_dependency(self, package: str, top_levels: List[str]) -> bool:
+    def _has_matching_dependency(self, package: Optional[str], top_levels: List[str]) -> bool:
         """
         Check if this module is provided by a listed dependency. This is the case if either the package name that was found in the metadata is
         listed as a dependency, or if the we found a top-level module name match earlier.
         """
-        return (package in [dep.name for dep in self.dependencies]) or len(top_levels) > 0
+        return package and (package in [dep.name for dep in self.dependencies]) or len(top_levels) > 0
 
-    def _has_matching_dev_dependency(self, package: str, dev_top_levels: List[str]) -> bool:
+    def _has_matching_dev_dependency(self, package: Optional[str], dev_top_levels: List[str]) -> bool:
         """
         Same as _has_matching_dependency, but for development dependencies.
         """
-        return (package in [dep.name for dep in self.dev_dependencies]) or len(dev_top_levels) > 0
+        return package and (package in [dep.name for dep in self.dev_dependencies]) or len(dev_top_levels) > 0

--- a/deptry/module.py
+++ b/deptry/module.py
@@ -9,10 +9,8 @@ from isort.stdlibs.py38 import stdlib as stdlib38
 from isort.stdlibs.py39 import stdlib as stdlib39
 from isort.stdlibs.py310 import stdlib as stdlib310
 
+from deptry.compat import PackageNotFoundError, metadata
 from deptry.dependency import Dependency
-from deptry.utils import import_importlib_metadata
-
-metadata, PackageNotFoundError = import_importlib_metadata()
 
 
 class Module:

--- a/deptry/result_logger.py
+++ b/deptry/result_logger.py
@@ -7,7 +7,7 @@ class ResultLogger:
     Display the issues to the user, and return exit-status 0 or 1 depending on if any issues were found.
     """
 
-    def __init__(self, issues: Dict[str, List[str]]):
+    def __init__(self, issues: Dict[str, List[str]]) -> None:
         self.issues = issues
 
     def log_and_exit(self) -> None:
@@ -25,7 +25,7 @@ class ResultLogger:
             self._log_additional_info()
 
     @staticmethod
-    def _log_total_number_of_issues_found(number):
+    def _log_total_number_of_issues_found(number: int) -> None:
         if number == 0:
             logging.info("Success! No dependency issues found.")
         elif number == 1:

--- a/deptry/utils.py
+++ b/deptry/utils.py
@@ -2,7 +2,7 @@ import os
 import sys
 from contextlib import contextmanager
 from pathlib import Path
-from typing import Dict
+from typing import Dict, Generator
 
 if sys.version_info >= (3, 11):
     import tomllib
@@ -13,7 +13,7 @@ PYPROJECT_TOML_PATH = "./pyproject.toml"
 
 
 @contextmanager
-def run_within_dir(path: Path) -> None:
+def run_within_dir(path: Path) -> Generator[None, None, None]:
     """
     Utility function to run some code within a directory, and change back to the current directory afterwards.
 

--- a/deptry/utils.py
+++ b/deptry/utils.py
@@ -1,9 +1,8 @@
 import os
 import sys
-import types
 from contextlib import contextmanager
 from pathlib import Path
-from typing import Dict, Tuple
+from typing import Dict
 
 if sys.version_info >= (3, 11):
     import tomllib
@@ -32,22 +31,6 @@ def run_within_dir(path: Path) -> None:
         yield
     finally:
         os.chdir(oldpwd)
-
-
-def import_importlib_metadata() -> Tuple[types.ModuleType, Exception]:
-    """
-    importlib.metadata is in the standard library since Python version 3.8
-    """
-    if sys.version_info[1] == 7:
-        import importlib_metadata as metadata
-        from importlib_metadata import PackageNotFoundError
-
-        return metadata, PackageNotFoundError
-    else:
-        import importlib.metadata as metadata
-        from importlib.metadata import PackageNotFoundError
-
-        return metadata, PackageNotFoundError
 
 
 def load_pyproject_toml(pyproject_toml_path: str = PYPROJECT_TOML_PATH) -> Dict:

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -52,6 +52,7 @@ exclude = '''
 profile = "black"
 
 [tool.mypy]
+files = ["deptry"]
 disallow_untyped_defs = true
 disallow_any_unimported = true
 no_implicit_optional = true
@@ -60,11 +61,6 @@ warn_return_any = false
 warn_unused_ignores = true
 show_error_codes = true
 ignore_missing_imports = true
-exclude = [
-    'venv',
-    '\.venv',
-    'tests'
-]
 
 [tool.deptry]
 exclude = [

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -52,14 +52,14 @@ exclude = '''
 profile = "black"
 
 [tool.mypy]
-disallow_untyped_defs = "True"
-disallow_any_unimported = "True"
-no_implicit_optional = "True"
-check_untyped_defs = "True"
-warn_return_any = "False"
-warn_unused_ignores = "True"
-show_error_codes = "True"
-ignore_missing_imports = "True"
+disallow_untyped_defs = true
+disallow_any_unimported = true
+no_implicit_optional = true
+check_untyped_defs = true
+warn_return_any = false
+warn_unused_ignores = true
+show_error_codes = true
+ignore_missing_imports = true
 exclude = [
     'venv',
     '\.venv',

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -19,7 +19,7 @@ importlib-metadata = { version = "*", python = "<=3.7" }
 chardet = "^5.0.0"
 
 [tool.poetry.dev-dependencies]
-mypy = "^0.961"
+mypy = "^0.981"
 pre-commit = "^2.20.0"
 pytest = "^7.1.2"
 pytest-cov = "^3.0.0"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -23,6 +23,7 @@ mypy = "^0.961"
 pre-commit = "^2.20.0"
 pytest = "^7.1.2"
 pytest-cov = "^3.0.0"
+types-chardet = "^5.0.4"
 
 [build-system]
 requires = ["poetry-core>=1.0.0"]

--- a/tests/test_dependency.py
+++ b/tests/test_dependency.py
@@ -1,7 +1,6 @@
 from unittest.mock import patch
 
 from deptry.dependency import Dependency
-from deptry.utils import import_importlib_metadata
 
 
 def test_simple_dependency():

--- a/tox.ini
+++ b/tox.ini
@@ -35,6 +35,7 @@ whitelist_externals = poetry
 commands =
     poetry install -v
     pytest --doctest-modules tests
+    mypy
 
 [pytest]
 norecursedirs = data*

